### PR TITLE
Multi wiggle template

### PR DESCRIPTION
**Pull request by cmdcolin**
This PR adds

- Track type for multibigwig
- Local/global autoscaling that works across multiple files

### CODE REVIEW

**Comment by cmdcolin** [comment]
Some next steps and or possible todo here include

- density type renderer
- y-scale bar display
- track config editor for multiple subtracks (currently crashes)

File: packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js [hunk 1]
```diff
@@ -27,6 +55,48 @@ export default class BigWigAdapter extends BaseAdapter {
...
+      })
+    }, blank)
+    return rectifyStats(stats)
+  }
+
+  // todo: add caching
+  async getLocalStats(regions = []) {
+    const ret = await Promise.all(regions.map(r => this.getRegionStats(r)))
+    const s = ret.reduce((a, b) => {
+      return Object.assign({}, a, {
```
Comment (rbuels) [inline packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js]: spread operator

File: packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js [hunk 2]
```diff
@@ -27,6 +55,48 @@ export default class BigWigAdapter extends BaseAdapter {
     return header.totalSummary
   }
 
+  // todo: add caching
+  // todo: incorporate summary blocks
+  async getRegionStats({ refName, start, end }) {
```
Comment (rbuels) [inline packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js]: @garrettjstevens we should make sure the stats-fetching is included in #192 adapter capabilities

File: packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js [hunk 3]
```diff
@@ -27,6 +55,48 @@ export default class BigWigAdapter extends BaseAdapter {
...
+      Math.ceil(end),
+    )
+    const blank = blankStats(end - start)
+    if (!feats.length) {
+      blank.scoreMax = 0
+      blank.scoreMin = 0
+    }
+    const stats = feats.reduce((a, b) => {
+      const s = b.score || 0
+      return Object.assign({}, a, {
```
Comment (rbuels) [inline packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js]: spread operator

File: packages/jbrowse-web/src/plugins/WiggleTrack/components/YScaleBar.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    filter: 'url(#dilate)',
+  },
+})
+const range = (start, stop, step) =>
+  Array.from({ length: (stop - start) / step + 1 }, (_, i) => start + i * step)
+
+function YScaleBar(props) {
+  const { min, max, classes } = props
+  return (
+    <Fragment key="refseq-label">
```
Comment (rbuels) [inline packages/jbrowse-web/src/plugins/WiggleTrack/components/YScaleBar.js]: JSX shorthand for a fragment can be just `<> </>`

File: packages/jbrowse-web/src/plugins/WiggleTrack/model.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      return {
+        min: Math.min(curr.scoreMin, acc.min),
+        max: Math.max(curr.scoreMax, acc.max),
+      }
+    },
+    { min: Infinity, max: -Infinity },
+  )
+}
+function finishLoading(stats) {
+  return Object.assign({}, stats, { loaded: true })
```
Comment (rbuels) [inline packages/jbrowse-web/src/plugins/WiggleTrack/model.js]:
```
return { ...stats, loaded: true }
```

File: packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js [hunk 6]
```diff
@@ -1,10 +1,38 @@
...
+  }
+  return variance < 0 ? 0 : Math.sqrt(variance)
+}
+
+function rectifyStats(s) {
+  return Object.assign({}, s, {
+    scoreMean: s.featureCount ? s.scoreSum / s.featureCount : 0,
+    scoreStdDev: calcStdFromSums(s.scoreSum, s.scoreSumSquares, s.featureCount),
+    featureDensity: s.featureCount / s.basesCovered,
+  })
```
Comment (rbuels) [inline packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js]:
```
return {
     ...s,
    scoreMean: s.featureCount ? s.scoreSum / s.featureCount : 0,
    scoreStdDev: calcStdFromSums(s.scoreSum, s.scoreSumSquares, s.featureCount),
    featureDensity: s.featureCount / s.basesCovered,
  })
```

File: packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js [hunk 7]
```diff
@@ -27,6 +55,48 @@ export default class BigWigAdapter extends BaseAdapter {
...
+    const s = ret.reduce((a, b) => {
+      return Object.assign({}, a, {
+        scoreMax: Math.max(a.scoreMax, b.scoreMax),
+        scoreMin: Math.min(a.scoreMin, b.scoreMin),
+        scoreSum: a.scoreSum + b.scoreSum,
+        scoreSumSquares: a.scoreSumSquares + b.scoreSumSquares,
+        featureCount: a.featureCount + b.featureCount,
+        basesCovered: a.basesCovered + b.basesCovered,
+      })
+    }, blankStats(0))
```
Comment (rbuels) [inline packages/jbrowse-web/src/plugins/BigWigAdapter/BigWigAdapter.js]: `...a`

**Comment by cmdcolin** [comment]
I added a couple features to this now including


a) allow setting min score and max score across all tracks
b) subtrack labels based on the name of the subtrack
c) mouseover scores
d) stabilizing how scale bars work api-wise

**Comment by rbuels** [comment]
few bugs in this branch I noticed just now while poking at it:
* the mouseover scores flicker, probably caused by a mouse-leave event getting fired when the mouse goes over the score marker. i remember having that bug in jbrowse 1 back in the day.
* the default height is too small, it should be at least big enough to show the track controls and track label.

**Comment by rbuels** [comment]
another bug I saw, switching to the volvox example dataset and then switching some tracks on and off causes a crash:

image

**Comment by cmdcolin** [comment]
I believe that switching datasets example is a result of this bug, I ran into it yesterday too https://github.com/GMOD/jbrowse-components/issues/219

**Comment by rbuels** [comment]
i guess try merging master and see if that crash goes away

**Comment by cmdcolin** [comment]
Looks like merging in master doesn't make that bug with changing datasets go away. The issue is specifically switching to the default volvox example causes issue. I think that switching between my custom "human" and "simplebw" datasets works ok

**Comment by cmdcolin** [comment]
Actually hold up, maybe I didn't merge master right

**Comment by cmdcolin** [comment]
Will close for now and reopen for single wiggle

**Pull request closed without merging**